File: scripts/algorithms/P/Preimage Size of Factorial Zeroes Function/Preimage Size of Factorial Zeroes Function.cpp
// Runtime: 5 ms (Top 11.74%) | Memory: 5.9 MB (Top 34.16%)
#define ll long long
class Solution {
public:
    bool check(ll num, int k){
        ll start = 5;
        ll ans = 0;
        while(start <= num){
            ans += (num/start);
            start *= 5;
        }
        return ans <= k;
    }
    ll get(int k){
        ll lo = 1;ll hi = 1e18;
        ll mid, res(1e18);
        while(lo <= hi){
            mid = (lo+hi)/2;
            if(check(mid, k)){
                lo = mid+1;
                res = mid;
            }
            else
                hi = mid-1;
        }
        return res;
    }
    int preimageSizeFZF(int k) {
        if(k == 0)
            return 5;
        return (get(k) - get(k-1));
    }
};

File: scripts/algorithms/P/Preimage Size of Factorial Zeroes Function/Preimage Size of Factorial Zeroes Function.py
// Runtime: 30 ms (Top 93.33%) | Memory: 16.50 MB (Top 82.96%)

class Solution:
    def preimageSizeFZF(self, k: int) -> int:  

        def atMost_k(k: int)-> int:

            left, right = 0, 5*k + 4

            while left <= right:
                mid = (left+right)//2
                count, n = 0, mid

                while n:
                    n//= 5
                    count+= n

                if count <= k: left = mid + 1
                else: right = mid - 1

            return right

        
        return atMost_k(k) - atMost_k(k-1)
